### DISCUSSION TITLE
Issue when creating a PDF with snakemake

### DISCUSSION THREAD

**Discussion by Sonia-Heye** [Deltares-research/FAIR-data-example-project#7] [Q&A] [answered]
Hey, 
In step **4.3 Finish your snakefile**, when trying to make the snakemake diagram by running: snakemake --dag | dot -Tpdf -o dag.pdf
I get the error: 
Building DAG of jobs...
Error: <stdin>: syntax error in line 1 near 'digraph'

Any idea where this is coming from and how I could fix it? It’s not critical, but it would be cool to see the .pdf generated

**Comment by sibrenloos** [comment]
I encounter the same error (no solution yet)

**Comment by JoerivanEngelen** [comment]
I think this has to do with line endings, on my machine using the -o option worked, so I hoped that it would fix this annoying issue once and for all. I know from experience that Snakemake/Graphviz do not support Windows that nicely when it comes to drawing graphs. I don't know an easy fix, but I can provide some workarounds:

# Workaround 1

Use:

```
snakemake --dag | dot -Tsvg > dag.svg
```

The generated .svg file can be viewed in the browser (e.g. Edge), and can be manually converted to a PDF by loading it in Inkscape and saving as PDF. 

# Workaround 2

The root of the issue lies with Linux and Windows having different line endings ( LF vs CRLF). Graphviz doens't seem to be able to deal with CRLF endings, outputted by Snakemake. Git for windows comes with a tool dos2unix, which helps converting this. 
You can access it by starting *"Git Bash"* instead of Powershell and running:
```
cd /path/to/folder/
pixi shell
snakemake --rulegraph | dos2unix | dot -Tpdf -o docs/dag.pdf
```

Alternatively, using Powershell, you can try this:

```
cd /path/to/folder/
pixi add git
pixi shell
snakemake --rulegraph | dos2unix | dot -Tpdf -o docs/dag.pdf
```

@Sonia-Heye & @sibrenloos, could you test if  one of these works?

**Reply by Sonia-Heye** [reply]
Hey Joeri, 
Thanks for your tips! Unfortunately all of the above solutions provide the same error for me as initially mentioned, except the Git Bash version, where I get a new error when running `pixi shell`: Unsupported shell: Bash(Bash)

**Reply by JoerivanEngelen** [reply]
Ah that's unfortunate, is the error already thrown when just calling ``snakemake --rulegraph``?

**Reply by Sonia-Heye** [reply]
No, calling just `snakemake --rulegraph` works. My output: 

`Building DAG of jobs...
digraph snakemake_dag {
    graph[bgcolor=white, margin=0];
    node[shape=box, style=rounded, fontname=sans,                 fontsize=10, penwidth=2];
    edge[penwidth=2, color=grey];
        0[label = "plot_heads", color = "0.60 0.6 0.85", style="rounded"];
        1[label = "post_process", color = "0.47 0.6 0.85", style="rounded"];
        2[label = "run_model", color = "0.00 0.6 0.85", style="rounded"];
        3[label = "build_model", color = "0.20 0.6 0.85", style="rounded"];
        4[label = "discretization", color = "0.07 0.6 0.85", style="rounded"];
        5[label = "download_data", color = "0.27 0.6 0.85", style="rounded"];
        6[label = "surface_water", color = "0.13 0.6 0.85", style="rounded"];
        7[label = "recharge", color = "0.40 0.6 0.85", style="rounded"];
        8[label = "initial_condition", color = "0.53 0.6 0.85", style="rounded"];
        9[label = "subsurface", color = "0.33 0.6 0.85", style="rounded"];
        1 -> 0
        2 -> 1
        3 -> 2
        4 -> 3
        6 -> 3
        9 -> 3
        7 -> 3
        8 -> 3
        5 -> 4
        5 -> 6
        5 -> 7
        4 -> 7
        5 -> 8
        4 -> 8
        5 -> 9
}`

**Reply by JoerivanEngelen** [reply]
Could be that ``dot`` has to be configured once. Could you try once:

```
pixi shell
dot -c
```

Then see if it works:

```
snakemake --dag | dot -Tpdf -o dag.pdf
```

**Reply by Sonia-Heye** [reply]
`dot -c` gives me the error:
`Error: failed to resolve gvplugin_kitty_LTX_library in ...\.pixi\envs\default\Library\bin\gvplugin_kitty.dll`
`Error: failed to resolve gvplugin_vt_LTX_library in ...\.pixi\envs\default\Library\bin\gvplugin_vt.dll`

Note: I have removed the first section of my folder directory above for simplicity. But due to these errors, the snakemake command is still not successful either.

**Comment by JoerivanEngelen** [comment, marked as answer]
I've reproduced the problem locally and figured out how to fix this. 

It is because of an encoding issue: Powershell by default encodes text slightly differently (UTF-16 BOM) than how a Unix tool like Graphviz (providing the ``dot`` tool) expects it (UTF-8). [See also this issue](https://stackoverflow.com/questions/40098771/changing-powershells-default-output-encoding-to-utf-8). Getting a nice fix for this requires Powershell 6.0, which requires extra installation, so I think too much effort for this tutorial. To work around the issue:

Run the following two commands:

```
snakemake --dag | out-file -encoding ASCII snakemake_graph.dot
dot -Tpdf -o dag.pdf snakemake_graph.dot
```

This encodes the text to ASCII, which is a subset of UTF-8 BOM-less. I'll update the tutorial as most people run into this.
Could you verify if this resolves your issue @Sonia-Heye, @sibrenloos?

UPDATE
---------
I have a verification of one user that this worked on their machine as well. Marking as answered. Feel free to unmark if it doesn't work for you.

**Reply by Sonia-Heye** [reply]
Thanks a lot and yes, this has resolved the issue.
